add Peierls creep rheology

naliboff opened the pull request:
This PR add a rheology module for a Peierls creep flow law, which is used in the `visco plastic` material model. The current formulation is an approximate solution for the Peierls creep rheology following a derivation from @mibillen.

While the current implementation returns the correct viscosity, a number of things need to be done before a full review:

1. Rechecking the implemented formulation against the analytical solution with @mibillen 
2. Add solution files for the provided test
3. Adding an option to use the full generalized Peierls creep flow law, which uses stress, rather than strain rate, as input. This would require a proper decomposition of the stress between different flow law mechanisms, as is done in `diffusion_dislocation`.
4. Add in the option to account different phases, following #3598.
 
I've opened this PR so @mibillen, @danieldouglas92, and others can take a look, add comments and make suggestions.

* [X] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).
* [X] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

bobmyhill commented:
Are there docs for the derivation? I'd like to see how the approximation is done, and how good the approximation is at different values of sigma/sigma* :) 
If this is only an approximate implementation of Peierls, perhaps calling it PeierlsApprox would be clearer? Or, if the fitting parameter can be approximated internally, adding a switch for the precise calculation with an Assert (precise_calculation==false, ExcNotImplemented("Precise Peierls not yet implemented"));

naliboff commented:
@bobmyhill - Thanks for the initial comments! Here is a link to the notes and scripts developed by @mibillen: https://magalibillen.faculty.ucdavis.edu/geodynamics-jupyter-notebooks/.

I agree we should making it very clear this is an approximation to the Peierls creep viscosity. I think the default should be the precise formulation, and then the approximated value option can be used if the iterative portioning scheme between diffusion, dislocation, and Peierls creep is not performed. I'll add in the precise formulation next.

naliboff commented:
This PR is ready for a formal review and as always comments are welcome from all parties. 

@bobmyhill - I've attempted to make it clear that the current formulation is an approximation, rather than the precise formulation. @mibillen - I've included you in the change log, as this was all derived from your work.

bobmyhill commented on tests/visco_plastic_peierls_mei_viscosity_approx.prm:
```diff
@@ -0,0 +1,118 @@
+#
+#  Fitting parameters for approximated viscosity:
+#    q = 0.5
+#    g = 0.1
```
But g is also gravitational acceleration (see above)? Maybe a different parameter name would avoid confusion?

naliboff replied:
will switch `g` to `gamma`

bobmyhill commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,169 @@
+                               peierls_fitting_exponent * std::pow(peierls_fitting_parameter,peierls_fitting_exponent);
+
+              const double left_term = 0.5 * (peierls_fitting_parameter * peierls_stress) /
+                                       std::pow((prefactors_peierls[composition] * std::pow(peierls_fitting_parameter * peierls_stress,2)),( 1 / (s + 2)));
```
looking at Magali's derivation, it doesn't look like $n$ needs to be a fixed exponent (the powers of two in the viscosity expression). The Raterron et al. (2004) expression for Peierls creep in olivine has n=0, and it would be nice to also use this derivation to approximate their expression.

bobmyhill replied:
It might also be good to follow other material models in ASPECT and include an activation volume for Peierls creep ("Eact" -> "Eact + P*Vact")

naliboff replied:
I'll go back and check if we can use a general value of `n`. @mibillen - Thoughts on both comments above?

mibillen replied:
I'll look at this today. Originally, I only derived this for the specific values of n, q, and p for the Mei flow law.

bobmyhill replied:
Certainly +P*Vact can be added without any problems. 
The derivation at the moment is only applicable to the case p = 1, but there might be a clever way to get round this.

naliboff commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,107 @@
+           * Currently, the only available option directly calculates the viscosity
+           * using an approximation to the model of Mei et al., 2010. The full
+           * derivation for this approximation (derived by Magali Billen) can be
+           * found at https://ucdavis.app.box.com/s/scaorcblr9u294836pgk4hyf60gv7psr.
```
Yes, I think we absolutely need to get this into the manual. The question is before this PR is merged or after, in conjunction with the summary @cedrict has been working on. I think it makes sense to a complete description of all the material behavior first, rather than just the new Peierls derivation + description.

naliboff commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,169 @@
+               * R is the gas constant
+               */
+              const double s = (activation_energies_peierls[composition] / (constants::gas_constant * temperature)) *
+                               peierls_fitting_exponent * std::pow(peierls_fitting_parameter,peierls_fitting_exponent);
```
Good point, I will make these all composition-dependent.

mibillen commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,169 @@
+               * An approximation of the Peierls creep formulation from Mei et al. (2010), where stress is
+               * replaced with strain rate (second invariant) when calculation viscosity. This substitution
+               * require two additional fitting parameters (g,q). Details of this derivation can be found
+               * in the following document: https://ucdavis.app.box.com/s/scaorcblr9u294836pgk4hyf60gv7psr.
```
I have the explanation in latex, so I can turn this into an appropriate explanation for the manual (fewer math steps shown for derivation).

naliboff commented:
I've addressed most of the comments from @bobmyhill, aside from:
1. Adding an activation volume
2. Moving documentation from the external reference to the the manual
3. Deriving a more general form (variable `n`, etc)

@bobmyhill - Would you have any objections to merging with the 3 items above left unchecked? My reasoning is this: some groups would like to start production models with Peierls creep and what is currently implemented works, albeit for a specific formulation. We can then tackle the above items piece by piece.

bobmyhill commented:
Hi @naliboff ... I have no problem with this. Maybe raise a new issue to fix the above?

naliboff commented:
@bobmyhill - yep, if the decision is to merge near this state then I'll open an issue outlining all the remaining pieces. Everyone involved is interested in using the precise formulation soon, so I don't see this being left by the wayside.

naliboff commented:
@bobmyhill @mibillen - The last commit updates the equations to use the general formulation. The code can definitely use another thorough review, especially for readability, naming of parameters, etc. I'll go through the code again later tonight and do some another round of cleanup. 

The existing test is now actually returning a viscosity slightly off (~ 5.36e18 Pa s) versus the analytical (~ 5.23e18 Pa s) solution. This will need to be debugged.

The formula itself is also quite hard to debug when using the full parameter names. Should we consider replacing the full names in the equations with simple variable names?.  Example:
`E = peierls_creep_activation_energies[c]`

gassmoeller commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,124 @@
+          /**
+           * List of Peirls creep prefactors (A).
+           */
+          std::vector<double> prefactors_peierls;
```
I would remove the `peierls` addition from all variables. It just make the names longer, and we are inside the peierls creep module, so no need for that.

naliboff replied:
Done and with other changes, things are indeed now more readable :)

gassmoeller commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,124 @@
+
+    namespace Rheology
+    {
+      template <int dim>
```
This needs an explanation about what Peierls creep is and where to read more about it.

naliboff replied:
I've added in @mibillen suggestion from below.

gassmoeller commented on include/aspect/material_model/visco_plastic.h:
```diff
@@ -382,6 +383,16 @@ namespace aspect
+        /**
+         * Whether to include peierls creep in the constitutive formulation.
+         */
+        bool use_peierls_creep;
```
Why is this not another viscosityscheme? Is it because it acts in addition to the other schemes? But is `composite` not also a composite scheme? And what happens for `composite` with added peierls creep? In that case the strain rate  should be distributed between the three different creep mechanisms I think. Maybe you discussed all that, but it is not obvious to me from the code alone.

naliboff replied:
I did things in this manner to ensure backwards compatibility. If we had Peierls creep as one of the viscosity schemes, then we would then need to include in the compositing with diffusion and dislocation creep. While we could set the default values to something that gives a very high viscosity (ex: 1e50 Pa s) and thus barely changes the composite viscosity values, I still thought it would be better to have it separately. 

Indeed, I checked for a simple example that doing the composite of Peierls creep with a composite of diffusion/dislocation creep gives the same thing as compositing all three values together. However, I will make derive the general expressions directly to be sure.

Having the Peierls creep separate also always composting with the Frank Kamenetskii viscosity, rather than diffusion/dislocation.

However, I'm open to restructuring in another way if it makes sense to do so. Restructuring will need to occur when we do the proper compositing as in `diffusion_dislocation.cc`, but for now I thought it would be best to ensure backwards compatibility.

bobmyhill replied:
I don't think any restructuring will be necessary for the compositing. We'll just need to add a strain_rate_and_derivatives function.

See #3759 for how I envisage this will work.

mibillen replied:
I did the math to check:
"Indeed, I checked for a simple example that doing the composite of Peierls creep with a composite of diffusion/dislocation creep gives the same thing as compositing all three values together. However, I will derive the general expressions directly to be sure."
Yes, it works out exactly the same.

gassmoeller replied:
Thanks for checking @mibillen! Then I am ok with adding it as it is done here, it is easier because this way it can be combined with any of the other viscosity schemes.

gassmoeller commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,229 @@
+               * T is temperature (K)
+               * R is the gas constant
+               */
+              const double s = ( (activation_energies_peierls[c] + pressure * activation_volumes_peierls[c]) / (constants::gas_constant * temperature)) *
```
If you remove all `peierls` components from the names below it will be much easier to read ;-). Also maybe there are slightly better names than `left`, `right` and so on. Coarsely it seems `left_term`  is a `stress_term`, `middle_term` is a `arrhenius_term`, and `right_term` is a `strain_rate_term`? Maybe there are better names. Or you could replace them by variable names in the formula above and use the variable names below?

naliboff replied:
I've adopted the naming schemes proposed above, definitely an improvement when reading through 👍 The documentation also now uses the conventions when writing out the formulation.

gassmoeller commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,229 @@
+                           "Select what type of Peierls creep flow law to use. Currently, the "
+                           "only available option is an approximation to Peierls creep, which "
+                           "uses the strain rate invariant, rather than stress, as an input. "
+                           "Future options will allow formulations that use stress as an input.");
```
Never mention future things in code ... you dont know if it will happen next week, next year or never and in all cases the mention is not helpful at the moment, except to appease reviewers who would complain about a parameter that has only a single option ;-).

naliboff replied:
Removed :-) I'll just open an issue as a reminder.

gassmoeller commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,229 @@
+                           "only available option is an approximation to Peierls creep, which "
+                           "uses the strain rate invariant, rather than stress, as an input. "
+                           "Future options will allow formulations that use stress as an input.");
+        prm.declare_entry ("Prefactors for Peierls creep", "1.4e-7",
```
Oh, is all of this added to the general `visco plastic` subsection and that is the reason it has the `Peierls` addition? Then keep `peierls` in the input file parameter, but drop it from the internal variables. Actually I would prefer if we could put all of this in a `Peierls creep` subsection. We did not do that for DiffusionCreep and DislocationCreep, because they were already around and we would have broken compatibility. Maybe we should also do the same for Frank Kamenetskii, but I did not think of it at the time.

naliboff replied:
Yes, it would make sense to do have subsections for all of the rheology modules. Should I do that now for Peierls creep, or do all of the rheology modules (DiffusionCreep, DislocationCreep, PeierlsCreep, FK, DruckerPrager, StrainDependent, etc) in a follow-up PR?

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -263,7 +263,13 @@ namespace aspect
             }
 
-          // Step 1d: multiply the viscosity by a constant (default value is 1)
+          // Step 1d-1: compute viscosity from Peierls creep law
+          const double viscosity_peierls = peierls_creep.compute_viscosity(edot_ii, pressure, temperature_for_viscosity, j);
+
+          // Step 1d-2: harmonically average diffusion/dislocation/composite and peierls creep viscosity
+          viscosity_pre_yield = (viscosity_pre_yield * viscosity_peierls) / (viscosity_pre_yield + viscosity_peierls);
+
```
This should be behind a `if (use_peierls_creep)`. Also like I asked above: Is this actually correct for the composite viscosity? Shouldnt be Peierls creep added in line 255? Or is this mathematically equivalent?

naliboff replied:
Doh, that was the whole point of adding that parameter.

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -795,6 +801,13 @@ namespace aspect
 
           // Frank-Kamenetskii viscosity parameters
           Rheology::FrankKamenetskii<dim>::declare_parameters(prm);
+          // Peierls creep parameters
```
empty line above this

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -929,6 +942,14 @@ namespace aspect
           frank_kamenetskii_rheology.initialize_simulator (this->get_simulator());
           frank_kamenetskii_rheology.parse_parameters(prm);
 
+          // Peierls creep parameters
+          use_peierls_creep = prm.get_bool ("Include Peierls creep");
+          if (use_peierls_creep)
+            {
+              peierls_creep.initialize_simulator (this->get_simulator());
+              peierls_creep.parse_parameters(prm);
+            }
```
Can you use the same structure of a unique pointer as for Frank Kamenetskii in #3766?

naliboff replied:
Actually, I don't think there is any need to use this check here and have removed it. `use_peierls_creep` is really only needed to decide about the compositing, which of course I forgot to add.

mibillen commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,124 @@
+           * Enumeration for selecting which type of Peierls creep flow law to use.
+           * Currently, the only available option directly calculates the viscosity
+           * using an approximation where the strain rate, rather than stress is used.
+           * This approximation requires specifying multiple fitting parameters. The full
```
there is only on fitting parameter for the approximation, gamma.p and q are no more fitting parameters of the original flow law than n is. p and q depend on the kind a crystallographic obstacles that resist dislocation glide. They don't have set names like n is the stress exponent, but maybe we could call them the glide_parameter_p and glide_parameter_q.

mibillen commented on tests/visco_plastic_peierls_mei_viscosity_approx.prm:
```diff
@@ -0,0 +1,125 @@
+#    s =  p*q*np.power(g,p)*np.power((1-np.power(g,p)),q-1)*(E/(R*T)) 
+#
+#  # Approximated viscosity
+#    visca = 0.5*gamma*sigp/np.power(A*np.power(gamma*sigp,n),1/(s+n))*np.power(edot_ii,(1/(s+n))-1)*np.exp((E/(R*T))*np.power((1 - np.power(gamma,p)),q)/(s+n)) 
+#          = 5.236440783251679e+18
```
When I calculate this in my script and plugging in the same numbers for T, P, V, and edot, I get visc   = 5.360939941500902e+18 
which matches what the Aspect puts out, check values that you put into the python script.

naliboff replied:
I get the same thing as well now. I forgot to calculate and update strain-rate for the approximation.

naliboff commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,169 @@
+               * edot_ii is the second invariant of the deviatoric strain rate
+               * T is temperature (K)
+               * R is the gas constant
+               */
```
I think all the parameters are defined now.

naliboff commented:
@gassmoeller @mibillen @bobmyhill - Thanks for the reviews and contributions! I believe I've addressed most of the comments, but there are a few additional discussion points left to resolve from the review by @gassmoeller.

naliboff commented:
After the last commit the testers are now happy 👍.

tjhei commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,136 @@
+           */
+          enum PeierlsCreepScheme
+          {
+            viscosity_approx
```
we typically spell out variables, maybe ``viscosity_approximation``?

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+              const double s = ( (activation_energies[c] + pressure * activation_volumes[c]) / (constants::gas_constant * temperature)) *
+                               glide_parameters_p[c] * glide_parameters_q[c] *
+                               std::pow((1. - std::pow(fitting_parameters[c], glide_parameters_p[c])),(glide_parameters_q[c] - 1.)) *
+                               std::pow(fitting_parameters[c], glide_parameters_p[c]);
```
holy cow.

naliboff replied:
Yep :-)

gassmoeller commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,136 @@
+          } peierls_creep_flow_law;
+
+          /**
+           * List of Peirls creep prefactors (A).
```
```suggestion
           * List of Peierls creep prefactors (A).
```

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "only available option is an approximation to Peierls creep, which "
+                           "uses the strain rate invariant, rather than stress, as an input. ");
+        prm.declare_entry ("Prefactors for Peierls creep", "1.4e-19",
+                           Patterns::List(Patterns::Double(0)),
```
```suggestion
                           Patterns::List(Patterns::Double(0.)),
```

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "If only one value is given, then all use the same value. "
+                           "Units: \\si{\\pascal}$^{-n_{\\text{peierls}}}$ \\si{\\per\\second}");
+        prm.declare_entry ("Stress exponents for Peierls creep", "2.0",
+                           Patterns::Anything(),
```
can this be a List(Double()) as well?

naliboff replied:
Thanks, I must have copied that from `dislocation.cc` and forgot to change it.

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "fields, for a total of N+1 values, where N is the number of compositional fields. "
+                           "If only one value is given, then all use the same value.  Units: None.");
+        prm.declare_entry ("Activation energies for Peierls creep", "320e3",
+                           Patterns::List(Patterns::Double(0)),
```
```suggestion
                           Patterns::List(Patterns::Double(0.)),
```

gassmoeller replied:
I think this comment has not been addressed yet.

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "for a total of N+1 values, where N is the number of compositional fields. "
+                           "If only one value is given, then all use the same value. Units: \\si{\\joule\\per\\mole}.");
+        prm.declare_entry ("Activation volumes for Peierls creep", "1.4e-5",
+                           Patterns::Anything(),
```
like above, can this be a list?

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "If only one value is given, then all use the same value. "
+                           "Units: \\si{\\meter\\cubed\\per\\mole}.");
+        prm.declare_entry ("Peierls stresses", "5.e9",
+                           Patterns::List(Patterns::Double(0)),
```
```suggestion
                           Patterns::List(Patterns::Double(0.)),
```

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "of compositional fields. If only one value is given, then all use the same value. "
+                           "Units: \\si{\\pascal}");
+        prm.declare_entry ("Peierls fitting parameters", "0.17",
+                           Patterns::List(Patterns::Double(0)),
```
```suggestion
                           Patterns::List(Patterns::Double(0.)),
```

gassmoeller commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,136 @@
+          std::vector<double> activation_volumes;
+
+          /**
+           * List of Peierls stresses (sigma_p)
+           */
+          std::vector<double> peierls_stresses;
+
+          /**
+           * List of Peierls fitting parameters (gamma)
+           */
+          std::vector<double> fitting_parameters;
+
+          /**
+           * List of the first Peierls parameter related
+           * to dislocation glide (p)
+           */
+          std::vector<double> glide_parameters_p;
+
+          /**
+           * List of the second Peierls parameter related
+           * to dislocation glide (q)
```
Since I am commenting anyway, please add periods at the end of documentation sentences above (or remove them uniformly, it is a matter of taste, but I am in the 'Comments are complete sentences too' fraction of the developer base :-)).

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "for a total of N+1 values, where N is the number of compositional fields. If only one "
+                           "value is given, then all use the same value. Units: none");
+        prm.declare_entry ("Peierls glide parameters p", "0.5",
+                           Patterns::List(Patterns::Double(0)),
```
```suggestion
                           Patterns::List(Patterns::Double(0.)),
```

tjhei commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -0,0 +1,231 @@
+                           "fields for a total of N+1 values, where N is the number of compositional fields. "
+                           "If only one value is given, then all use the same value. Units: none");
+        prm.declare_entry ("Peierls glide parameters q", "1.0",
+                           Patterns::List(Patterns::Double(0)),
```
```suggestion
                           Patterns::List(Patterns::Double(0.)),
```

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -263,7 +263,12 @@ namespace aspect
 
-          // Step 1d: multiply the viscosity by a constant (default value is 1)
+          // Step 1d: compute viscosity from Peierls creep law and harmonically average with current viscosities
+          const double viscosity_peierls = peierls_creep.compute_viscosity(edot_ii, pressure, temperature_for_viscosity, j);
```
It looks like you compute it even if it is not used, is that on purpose?

naliboff replied:
Yep, this should go inside if (use_peierls_creep). I initialize viscosity_peierls as an NaN right before.

gassmoeller commented on include/aspect/material_model/visco_plastic.h:
```diff
@@ -382,6 +383,16 @@ namespace aspect
+        /**
+         * Objects for computing peierls creep viscosities.
+         */
+        Rheology::PeierlsCreep<dim> peierls_creep;
```
Can you still make this into a unique_ptr as you did for Frank Kamenetskii?

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -263,7 +263,12 @@ namespace aspect
 
-          // Step 1d: multiply the viscosity by a constant (default value is 1)
+          // Step 1d: compute viscosity from Peierls creep law and harmonically average with current viscosities
+          const double viscosity_peierls = peierls_creep.compute_viscosity(edot_ii, pressure, temperature_for_viscosity, j);
```
this line should be inside `use_peierls_creep` as well, shouldnt it?

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -796,6 +801,14 @@ namespace aspect
+          // Peierls creep parameters
+          Rheology::PeierlsCreep<dim>::declare_parameters(prm);
+
+          // Whether to include Peierls creep in the rheological formulation
```
this comment is unnecessary (it is repeated two lines below).

naliboff commented:
/rebuild

naliboff commented:
@gassmoeller @tjhei - Thanks for the additional reviews! I think I've addressed all comments, but the testers are producing a very weird compiler error at the moment. Everything compiles fine on my end.

Tester compiler error:
`/home/runner/work/aspect/aspect/source/material_model/visco_plastic.cc:274:77: error: ‘pressure’ was not declared in this scope
208
               viscosity_peierls = peierls_creep->compute_viscosity(edot_ii, pressure, temperature_for_viscosity, j);`

The listed line number (`visco_plastic.cc:274:77`  for the line above also does not match with what is in the file. Is something odd going on with the tester?

tjhei commented:
> ut the testers are producing a very weird compiler error at the moment

I restarted the github actions testers and Jenkins is still behind (8 or so builds), so it should get to it soon.

tjhei commented:
Oh, now I get it: we are testing the PR merged into master and there have been other changes to visco_plastic that break your PR. @naliboff can you rebase to the current master, please?

naliboff commented:
@tjhei - Ahh, weird that there is not a merge conflict notice. I actually just tried to rebase and got the following error:
```
error: Your local changes to the following files would be overwritten by merge:
	doc/manual/cookbooks/burnman
	doc/manual/cookbooks/continental_extension
	doc/manual/cookbooks/crustal_deformation

```
I'll sort this out and then push the rebased branch.

tjhei commented:
Just delete those files/folders, reset all changes if you deleted files under version control, and try again.

anne-glerum commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -0,0 +1,136 @@
+          std::vector<double> prefactors;
+
+          /**
+           * List of Peierls creep stress exponents n.
```
Some symbols are within brackets, some are not. Nitpicky, but maybe unify?

naliboff replied:
not nitpicky at all, thanks for catching that!

naliboff commented:
@gassmoeller @tjhei - Rebased and I think all comments have been addressed. Testers are happy as well. Let me know if you would like any more changes!

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -267,7 +267,15 @@ namespace aspect
 
-          // Step 1d: multiply the viscosity by a constant (default value is 1)
+          // Step 1d: compute viscosity from Peierls creep law and harmonically average with current viscosities
+          double viscosity_peierls = numbers::signaling_nan<double>();
```
You do not need this outside the `if`. Make line 274 into `const double viscosity_peierls = ...`

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -946,6 +961,12 @@ namespace aspect
 
+          // Peierls creep parameters
+          use_peierls_creep = prm.get_bool ("Include Peierls creep");
+          peierls_creep = std_cxx14::make_unique<Rheology::PeierlsCreep<dim>>();
+          peierls_creep->initialize_simulator (this->get_simulator());
+          peierls_creep->parse_parameters(prm);
```
You want to put these three lines into a `if (use_peierls_creep)`

bobmyhill commented on tests/visco_plastic_peierls_mei_viscosity_approx.prm:
```diff
@@ -0,0 +1,130 @@
+# A test for the peierls creep viscositiy approximation  
```
As Rene left a couple of additional things, I also noticed this:
viscositiy -> viscosity
